### ISSUE TITLE
Crate + client version 0.54.6 throw exception with complex primary key

### ISSUE THREAD

**Issue by hasnainjaved** [crate/crate#3320]
Hello,

When inserting a document with a complex primary key and subsequently querying the document using it's complex primary results in an exception.

    CREATE TABLE EntityWithComplexId (
        "entity_class" string,
        "type" string,
        "complexId" object AS (
            "stringField" string,
            "booleanField" boolean,
            "levelZero" object AS (
                "levelZeroIntField" integer,
                "levelZeroLongField" long,
                "levelOne" object AS (
                    "levelOneDoubleField" double
                )
            )
        ) PRIMARY KEY
    ) WITH (number_of_replicas='0', refresh_interval='1000', column_policy='dynamic');

    INSERT INTO EntityWithComplexId ("complexId","entity_class","type") VALUES ({booleanField=true, levelZero={levelOne={levelOneDoubleField=1.0}, levelZeroIntField=5, levelZeroLongField=4}, stringField='CRATE'}, 'com.xxx.yyy.EntityWithComplexId', 'complex');

    SELECT * FROM EntityWithComplexId WHERE "complexId" = {booleanField=true, levelZero={levelOne={levelOneDoubleField=1.0}, levelZeroIntField=5, levelZeroLongField=4}, stringField='CRATE'};

Error message using the above select query

    SQLActionException[cannot cast {levelzero={levelone={levelonedoublefield=1.0}, levelzerolongfield=4, levelzerointfield=5}, booleanfield=true, stringfield=CRATE} to string]

Enclosing the complex key in double quotes does not work either 

    SQLActionException[Column {booleanField=true, levelZero={levelOne={levelOneDoubleField=1.0}, levelZeroIntField=5, levelZeroLongField=4}, stringField='CRATE'} unknown]

Crate client Stacktrace below

```
at io.crate.action.sql.TransportBaseSQLAction.esToCrateException(TransportBaseSQLAction.java:313)
	at io.crate.action.sql.TransportBaseSQLAction.buildSQLActionException(TransportBaseSQLAction.java:358)
	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:206)
	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:191)
	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:80)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:75)
	at io.crate.action.sql.TransportSQLAction$TransportHandler.messageReceived(TransportSQLAction.java:139)
	at io.crate.action.sql.TransportSQLAction$TransportHandler.messageReceived(TransportSQLAction.java:127)
	at org.elasticsearch.transport.netty.MessageChannelHandler.handleRequest(MessageChannelHandler.java:222)
	at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:114)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.elasticsearch.common.netty.OpenChannelsHandler.handleUpstream(OpenChannelsHandler.java:96)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: cannot cast {stringField=CRATE, levelZero={levelZeroLongField=4, levelOne={levelOneDoubleField=1.0}, levelZeroIntField=5}, booleanField=true} to string
	at io.crate.types.StringType.value(StringType.java:74)
	at io.crate.analyze.symbol.ValueSymbolVisitor$3.visitLiteral(ValueSymbolVisitor.java:63)
	at io.crate.analyze.symbol.ValueSymbolVisitor$3.visitLiteral(ValueSymbolVisitor.java:60)
	at io.crate.analyze.symbol.Literal.accept(Literal.java:137)
	at io.crate.analyze.symbol.SymbolVisitor.process(SymbolVisitor.java:30)
	at io.crate.analyze.symbol.ValueSymbolVisitor.process(ValueSymbolVisitor.java:43)
	at io.crate.analyze.symbol.ValueSymbolVisitor$1.apply(ValueSymbolVisitor.java:38)
	at io.crate.analyze.symbol.ValueSymbolVisitor$1.apply(ValueSymbolVisitor.java:34)
	at com.google.common.collect.Lists$TransformingRandomAccessList$1.transform(Lists.java:617)
	at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48)
	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:302)
	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:289)
	at io.crate.analyze.Id$2.apply(Id.java:55)
	at io.crate.analyze.Id$2.apply(Id.java:50)
	at io.crate.analyze.where.DocKeys$DocKey.id(DocKeys.java:88)
	at io.crate.executor.transport.task.elasticsearch.ESGetTask$GetResponseContext.<init>(ESGetTask.java:307)
	at io.crate.executor.transport.task.elasticsearch.ESGetTask.<init>(ESGetTask.java:80)
	at io.crate.executor.transport.TransportExecutor$NodeVisitor.visitESGetNode(TransportExecutor.java:250)
	at io.crate.executor.transport.TransportExecutor$NodeVisitor.visitESGetNode(TransportExecutor.java:237)
	at io.crate.planner.node.dql.ESGetNode.accept(ESGetNode.java:92)
	at io.crate.executor.transport.TransportExecutor$TaskCollectingVisitor.visitIterablePlan(TransportExecutor.java:161)
	at io.crate.executor.transport.TransportExecutor$TaskCollectingVisitor.visitIterablePlan(TransportExecutor.java:155)
	at io.crate.planner.IterablePlan.accept(IterablePlan.java:59)
	at io.crate.planner.PlanVisitor.process(PlanVisitor.java:37)
	at io.crate.executor.transport.TransportExecutor.newJob(TransportExecutor.java:128)
	at io.crate.action.sql.TransportBaseSQLAction.executePlan(TransportBaseSQLAction.java:234)
	at io.crate.action.sql.TransportBaseSQLAction.processAnalysis(TransportBaseSQLAction.java:225)
	at io.crate.action.sql.TransportBaseSQLAction.doExecute(TransportBaseSQLAction.java:203)
	... 32 more
```

**Comment by joemoe** [comment]
hi @hasnainjaved, we don't support complex data types as primary keys. We will improve the error message as there's clearly some space for improvement. 

Can you tell me your use case, why you need to do this. Is this something that Spring needs?
Thanks for reporting.

**Comment by mfussenegger** [comment]
As a workaround you can can define each field of the object as primary key:

    CREATE TABLE EntityWithComplexId (
        "entity_class" string,
        "type" string,
        "complexId" object AS (
            "stringField" string primary key,
            "booleanField" boolean primary key,
            "levelZero" object AS (
                "levelZeroIntField" integer primary key,
                "levelZeroLongField" long primary key,
                "levelOne" object AS (
                    "levelOneDoubleField" double primary key
                )
            )
        )
    ) WITH (number_of_replicas='0', refresh_interval='1000', column_policy='dynamic');


But I've just found an issue with our object equality implementation in that it is too strict about numeric types.
In the case of your query

    SELECT * FROM EntityWithComplexId WHERE "complexId" = {booleanField=true, levelZero={levelOne={levelOneDoubleField=1.0}, levelZeroIntField=5, levelZeroLongField=4}, stringField='CRATE'};

It loads the `levelZeroLongField` as Integer (an optimization to reduce
serialization overhead) and then compares it with a 4L (Long) which causes the
query to not match anything.

We'll fix this in the next hotfix release.

**Comment by hasnainjaved** [comment]
Hi,

@joemoe Spring does not need this. I am one of the author's of Spring Data Crate and we wanted to bring Crate under the Spring Data family. I implemented the feature for supporting complex primary keys and it used to work with Crate version **0.47.4**. If Crate has dropped support for complex data types as primary keys, we will remove this feature from Spring Data Crate as well :)

@mfussenegger Thank your for the workaround but most probably we'll be removing support for complex objects as primary keys.

Kind Regards.

**Comment by jodok** [comment]
btw. - i met @kbastani - working with pivotal on spring yesterday - probably he can help as well.

**Comment by hasnainjaved** [comment]
@jodok We created a jira ticket for it last year

https://jira.spring.io/browse/DATACMNS-675

Would appreciate if this could be pushed forward and bring it to a GA release. Thanks.

Kind Regards.

**Comment by mfussenegger** [comment]
@hasnainjaved We weren't really aware that we *did* support primary keys on object columns. And I've just looked at the implementation in 0.47.4.
I think the implementation only appeared to work, but was also broken back then. (We just called .toString on a Map which can produce other results in Java 8 than it does on Java 7).

So although it appears to be a regression we've still decided to drop support for now. Because there is no way for us to fix the support in a backward compatible way.
